### PR TITLE
Allow to use Tooltips/Popovers in Shadow DOM

### PR DESCRIPTION
**Pull request by Johann-S**
Fixes : #24918

### CODE REVIEW

**Comment by Johann-S** [comment]
ping @XhmikosR

File: js/tests/visual/tooltip.html [hunk 1]
```diff
@@ -57,6 +62,21 @@ <h1>Tooltip <small>Bootstrap Visual Test</small></h1>
...
       $(function () {
+        if (typeof document.body.attachShadow === 'function') {
+          var shadowRoot = $('#shadow')[0].attachShadow({ mode: 'open' })
+          shadowRoot.innerHTML = [
+            '<button type="button" class="btn btn-secondary" data-toggle="tooltip" data-placement="top" title="Tooltip on top in a shadow dom">',
+              'Tooltip on top in a shadow dom',
+            '</button>',
+            '<button id="secondTooltip" type="button" class="btn btn-secondary" data-toggle="tooltip" data-placement="top" title="Tooltip on top in a shadow dom with container option">',
+              'Tooltip on top in a shadow dom',
+            '</button>'
```
Comment (XhmikosR) [inline js/tests/visual/tooltip.html]: Any reason you use an Array instead of a string?

**Comment by XhmikosR** [comment]
@Johann-S: I'm not familiar with shadow DOM at all, so I'm not sure I can be of any help...

File: js/tests/visual/tooltip.html (hunk 1 above)
Comment (Johann-S) [reply js/tests/visual/tooltip.html]: For me it's more readable but that's a matter of preference

**Comment by XhmikosR** [comment]
I'd rather we are consistent. Plus eventually we should use template
literals.

**Comment by Johann-S** [comment]
This is a live example : https://codepen.io/Johann-S/pen/ZvxgLZ

if @ericcarino or @LMOkino can take a look and see if it works as you wanted it would be perfect 👍

**Comment by LMOkino** [comment]
Johann, it looks good. Please let us know when it is in the bootstrap code.

**Comment by Johann-S** [comment]
@mdo if you have time to check that change, it seems it works pretty well

**Comment by mdo** [comment]
Is this intentional?

screen shot 2018-01-15 at 9 34 22 pm

**Comment by Johann-S** [comment]
Yep it's to show that if you use `container` option to append into the Shadow DOM you'll loose Bootstrap's style but if you include Bootstrap CSS in your Shadow DOM it should be fine

File: js/src/tooltip.js [hunk 2]
```diff
@@ -254,8 +254,9 @@ const Tooltip = (($) => {
       if (this.isWithContent() && this._isEnabled) {
         $(this.element).trigger(showEvent)
 
+        const shadowRoot = Util.findShadowRoot(this.element)
         const isInTheDom = $.contains(
-          this.element.ownerDocument.documentElement,
+          shadowRoot !== null ? shadowRoot : this.element.ownerDocument.documentElement,
```
Comment (thejohnfreeman) [inline js/src/tooltip.js]: This code is trying to find the "root" `Element`, either `document.documentElement` or a `ShadowRoot`, but it is split into two pieces: the `findShadowRoot` function, which may return `null`, and this ternary, which will handle `null`. Why not just fold these two pieces into one function whose purpose is to find the root element?

```js
// Returns the root Element, either document.documentElement or a ShadowRoot.
findRootElement (element) {
  // Go straight to the root node if the browser supports this method.
  if (typeof element.getRootNode === 'function') {
    const node = element.getRootNode()
    return (node instanceof ShadowRoot) ? node : node.documentElement
  }

  // Walk the parent chain to the root node.
  while (element.parentNode) {
    // Check for shadowRoot on the way up.
    if (element instanceof ShadowRoot) {
      return element
    }
    element = element.parentNode
  }
  // At this point, element has no parentNode. Thus, it is the root element.

  return element
}
```

Comment (Johann-S) [reply js/src/tooltip.js]: I agree that's better like the way you suggested 👍

**Comment by Johann-S** [comment]
@XhmikosR I rebased here, and made some changes, maybe we should add information in our docs ?

**Comment by XhmikosR** [comment]
Yup, we should mention it in docs. Can you add some basic info?

**Comment by Johann-S** [comment]
Done 👍

**Review by XhmikosR — APPROVED** [review]

**Pull request merged**